table/block_based/filter_policy_internal.h

```diff
@@ -196,28 +175,111 @@ class BloomFilterPolicy : public BuiltinFilterPolicy {
-class LevelThresholdFilterPolicy : public BuiltinFilterPolicy {
+// For NewBloomFilterPolicy
+//
+// This is a user-facing policy that automatically choose between
+// LegacyBloom and FastLocalBloom based on context at build time,
+// including compatibility with format_version.
+class BloomFilterPolicy : public BloomLikeFilterPolicy {
  public:
-  LevelThresholdFilterPolicy(std::unique_ptr<const FilterPolicy>&& a,
-                             std::unique_ptr<const FilterPolicy>&& b,
-                             int starting_level_for_b);
+  explicit BloomFilterPolicy(double bits_per_key);
 
+  // To use this function, call BuiltinFilterPolicy::GetBuilderFromContext().
+  //
+  // Neither the context nor any objects therein should be saved beyond
+  // the call to this function, unless it's shared_ptr.
   FilterBitsBuilder* GetBuilderWithContext(
-      const FilterBuildingContext& context) const override;
+      const FilterBuildingContext&) const override;
+
+  static const char* kName();
+  std::string GetId() const override;
+};
 
-  inline int TEST_GetStartingLevelForB() const { return starting_level_for_b_; }
+// For NewRibbonFilterPolicy
+//
+// This is a user-facing policy that chooses between Standard128Ribbon
+// and FastLocalBloom based on context at build time (LSM level and other
+// factors in extreme cases).
+class RibbonFilterPolicy : public BloomLikeFilterPolicy {
+ public:
+  explicit RibbonFilterPolicy(double bloom_equivalent_bits_per_key,
+                              int bloom_before_level);
 
-  inline const FilterPolicy* TEST_GetPolicyA() const { return policy_a_.get(); }
+  FilterBitsBuilder* GetBuilderWithContext(
+      const FilterBuildingContext&) const override;
 
-  inline const FilterPolicy* TEST_GetPolicyB() const { return policy_b_.get(); }
+  int GetBloomBeforeLevel() const { return bloom_before_level_; }
+
+  static const char* kName();
+  std::string GetId() const override;
 
  private:
-  const std::unique_ptr<const FilterPolicy> policy_a_;
-  const std::unique_ptr<const FilterPolicy> policy_b_;
-  int starting_level_for_b_;
+  const int bloom_before_level_;
 };
 
+// Deprecated block-based filter only. We still support reading old
+// block-based filters from any BuiltinFilterPolicy, but there is no public
+// option to build them. However, this class is used to build them for testing
+// and for a public backdoor to building them by constructing this policy from
+// a string.
+class DeprecatedBlockBasedBloomFilterPolicy : public BloomLikeFilterPolicy {
+ public:
+  explicit DeprecatedBlockBasedBloomFilterPolicy(double bits_per_key);
+
+  // Internal contract: returns a new fake builder that encodes bits per key
+  // into a special value from EstimateEntriesAdded(), using
+  // kSecretBitsPerKeyStart
+  FilterBitsBuilder* GetBuilderWithContext(
+      const FilterBuildingContext&) const override;
+  static constexpr size_t kSecretBitsPerKeyStart = 1234567890U;
+
+  static const char* kName();
```

Comment:
Is there a reason not to put the implementation here as well?  static const char* kName() { return "x.y.z"; }

Reply:
Functions defined in headers slow down compilation and linking, so I avoid them when not performance critical or templated. Is there a good reason for making these functions as requested or just "consistency"?